libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -67,5 +69,62 @@ bool PostTypeContractLevelChecker::check(ContractDefinition const& _contract)
 			errorHashes[hash][signature] = error->location();
 	}
 
+	if (auto const* layoutSpecifier = _contract.storageLayoutSpecifier())
+		checkStorageLayoutSpecifier(*layoutSpecifier);
+
 	return !Error::containsErrors(m_errorReporter.errors());
 }
+
+void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(StorageLayoutSpecifier const& _storageLayoutSpecifier)
+{
+	Expression const& baseSlotExpression = _storageLayoutSpecifier.baseSlotExpression();
+
+	if (!*baseSlotExpression.annotation().isPure)
+		// TODO: introduce and handle erc7201 as a builtin function
+		m_errorReporter.typeError(
+			1139_error,
+			baseSlotExpression.location(),
+			"The base slot of the storage layout must be a compile-time constant expression."
+		);
+
+	auto const* baseSlotExpressionType = type(baseSlotExpression);
+	if (auto const* rationalType = dynamic_cast<RationalNumberType const*>(baseSlotExpressionType))
+	{
+		std::string errorMessage = "Storage layout cannot be specified by ";
+		if (rationalType->isFractional())
+		{
+			errorMessage += "a fractional number.";
+			m_errorReporter.typeError(
+				1763_error,
+				baseSlotExpression.location(),
+				errorMessage
+			);
+		}
+		else if (
+			BoolResult convertibleToUint = baseSlotExpressionType->isImplicitlyConvertibleTo(*TypeProvider::uint256());
+			!convertibleToUint
+			)
+		{
+			errorMessage += "a number exceeding the range of type uint256. Current type is " + baseSlotExpressionType->humanReadableName();
+			errorMessage += ". " + convertibleToUint.message();
+			m_errorReporter.typeError(
+				6753_error,
+				baseSlotExpression.location(),
+				errorMessage
+			);
+		}
+		else
+		{
+			solAssert(!rationalType->isFractional());
+			solAssert(rationalType->value().denominator() == 1);
+			solAssert(rationalType->value().numerator() <= u256(-1));
+			_storageLayoutSpecifier.annotation().baseSlot = u256(rationalType->value().numerator());
+		}
+	}
+	else
+		m_errorReporter.typeError(
+			6396_error,
+			baseSlotExpression.location(),
+			"The storage layout can only be specified by number literals."
```

Comment:
```suggestion
			"The base slot of the storage layout must evaluate to a rational number."
```

Reply:
Also, not sure if there's much point in sharing the prefix of the error messages above. I'd simply repeat it in each messsage, it's not that long. It's generally easier to grep for messages in the code when they're in one chunk. You can use fmtlib to make it even less disjoint.